CHOPIN-48501: Reworked get_security_annotations script and updated README

bredy opened the pull request:
- using modification sequence number order on events ensuring
  all new/changes events are processed. Needed to change
  traversal order - now script is going from bottom (events) to
  top (alerts).

- taking customer ID from API, it's enough to relpace
  just SecureX parameter

bredy committed:
CHOPIN-48501: Reworked get_security_annotations script

- using modification sequence number order on events ensuring
  all new/changes events are processed. Needed to change
  traversal order - now script is going from bottom (events) to
  top (alerts).

- taking customer ID from API, it's enough to relpace
  just SecureX credential & visibility server parameters

- use bulk loading to get contextual referred objects

jimarek commented on get_security_annotations.py:
```diff
@@ -176,50 +211,45 @@ def log_event_attributes(alert, threat_intel_record, asset, threat_detection, ev
+            for threat_detection_id in threat_detection_ids:
+                # get parent objects to event - threat detections and alert
+                # and also threat intel record from threat-catalog bounded context
+                threat_detection_with_alert_ids = get_threat_detection_with_alert_ids(threat_detection_id)
```
It would be better to use bulk operations to get threat detections (also I would rather see to get all threat detections at once and not per event)

bredy replied:
Reworked to use bulk requests.

jimarek commented on get_security_annotations.py:
```diff
@@ -176,50 +211,45 @@ def log_event_attributes(alert, threat_intel_record, asset, threat_detection, ev
+                # get IDs references to alerts (usually one)
+                alert_ids = threat_detection_with_alert_ids["alertIds"]
+                for alert_id in alert_ids:
+                    alert = get_alert(alert_id)
```
The same note as for threat detections

bredy replied:
Reworked to use bulk requests.

cisco-jjudas commented on get_security_annotations.py:
```diff
@@ -6,60 +6,59 @@
-            previous_event_modified_at = parse_iso_datetime(previous_event_modified_at_file.readline())
-            previous_event_modified_at_file.close()
-            return previous_event_modified_at
+            events_end_cursor_file = open(EVENTS_END_CURSOR_FILENAME, "r")
```
you could use a with statement

bredy replied:
Used "with".

cisco-jjudas commented on get_security_annotations.py:
```diff
@@ -100,20 +101,13 @@ def __has_current_page_items(self):
-                            })
-    return response.json()
+def get_object(object_url):
+    global object_cache
```
it might be better to wrap all the functions in some kind of object, than you wouldn't need mutable global state

bredy replied:
Removed this code when rewriting it to bulk.

bredy committed:
CHOPIN-48501: Updated README according to script changes

mvelkCisco commented on README.md:
````diff
@@ -22,18 +22,28 @@ $ curl -X POST \
     "scope": "casebook"
 }
 ```
+In case something went wrong you'll get "400 Bad Request" with explanation in "error_description",
````
`In case something went wrong, you'll get "400 Bad Request" with an explanation in "error_description",`

mvelkCisco commented on README.md:
````diff
@@ -347,7 +357,7 @@ def get_all_alerts():
 ```
 
-> This script is available in the `get_all_alerts.py` file.
+> This script improved by direct usage of SecureX credential is available in the `get_all_alerts.py` file. 
````
`More complete version of this script is available in the `get_all_alerts.py` file.`

mvelkCisco commented on README.md:
```diff
@@ -362,97 +372,193 @@ To get the first page of `ThreatDetections`, use the following query:
+### Modification Sequence Number
 
-Most convicting `Events` contain `securityAnnotation`, which represents key observations used for detecting threat or malicious behavior.
+Because there are potentially a lots of `ThreatDetections` and magnitude more `Events` we need to have mechanism
```
These 2 paragraphs are not that great. Way too rambly for a documentation, many typos and bad grammar. I'll come back to these later with another proposal.

mvelkCisco committed:
CHOPIN-48501 updated readme

shcherbin commented on get_security_annotations.py:
```diff
@@ -176,50 +122,140 @@ def log_event_attributes(alert, threat_intel_record, asset, threat_detection, ev
+        write_events_end_cursor(events_iterator.end_cursor)
 
 
 main()
```
```suggestion
if __name__ == "__main__":
    main()
```

bredy replied:
ok

shcherbin commented on get_security_annotations.py:
```diff
@@ -1,172 +1,118 @@
             sys.exit(1)
     else:
-        return datetime.datetime.fromtimestamp(0)
+        return None
```
```suggestion
```

The `else` statement is redundant. No need to return `None` explicitly.

bredy replied:
ok

shcherbin commented on api_client.py:
```diff
@@ -38,3 +44,116 @@ def get_authorization(securex_host_name, secuerx_client_id, securex_client_passw
+    """
+    def __init__(self, collection_url_path, authorization_fn, query_params=None, cursor=None, request_body=None):
+        # keep immutable these constructor parameters
+        self.collection_url_path = collection_url_path
```
I might be nitpicking, but it's a good practice to define instance variables protected (i.e. defined with a leading underscore):

`self.query_params = query_params or {}` -> `self._query_params = query_params or {}`

bredy replied:
ok (for all)

shcherbin commented on api_client.py:
```diff
@@ -38,3 +44,116 @@ def get_authorization(securex_host_name, secuerx_client_id, securex_client_passw
+    return customer_id
+
+
+class CollectionIterator:
```
Optional suggestion: please consider implementing the iterable interface.

Refs:
- https://wiki.python.org/moin/Iterator
- Python 2 compatibility: https://stackoverflow.com/questions/29578469/how-to-make-an-object-both-a-python2-and-python3-iterator
- https://www.geeksforgeeks.org/python-__iter__-__next__-converting-object-iterator/

shcherbin commented on get_security_annotations.py:
```diff
@@ -176,50 +122,140 @@ def log_event_attributes(alert, threat_intel_record, asset, threat_detection, ev
+        cursor=previous_events_end_cursor
+    )
+    # remember events and get ids of related object (threat detections and assets) to "bulk load" them
     while events_iterator.has_next():
         event = events_iterator.next()
```
Once the iterator interface is implemented  the loop can be simplified as follows:

```suggestion
    for event in events_iterator:
```

shcherbin commented on api_client.py:
```diff
@@ -38,3 +44,116 @@ def get_authorization(securex_host_name, secuerx_client_id, securex_client_passw
+        }
+
+        if self.request_body:
+            response = requests.post(API_HOST_NAME + self.next_page_url, headers=headers, json=self.request_body)
```
consider using the session object: https://docs.python-requests.org/en/master/user/advanced/#session-objects

`requests.post|get` creates a new session on each method call.

bredy committed:
CHOPIN-48501: ApiClient class, GtaAuth class

- made Python code more pythonic based on review
- ApiClient is class now containig new GtaAuth to provide authorization in expected and reusable way

bredy committed:
CHOPIN-48501: Updated README according to scripts changes

mvelkCisco committed:
CHOPIN-48501 update readme

mvelkCisco committed:
CHOPIN-48501 remove GTA from comments
